### DISCUSSION TITLE
Wrong URL when fetching GitLab Changelogs

### DISCUSSION THREAD

**Discussion by MalteHei** [renovatebot/renovate#24509] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

We have a case where our app depends on a GitLab Pipeline which is located in the same GitLab instance.
Renovate can not get the changelogs for this dependency.

The cause seems to be that the URL to fetch is not being built correctly as our GitLab instance is only accessible via some path (see debug logs).

Reprocuding this may be difficult because I assume you need a dedicated GitLab instance...

### Relevant debug logs

<details><summary>config.js</summary>

```js
module.exports = {
	platform: 'gitlab',
	endpoint: 'https://gitlab.example.com/somePath/api/v4/',
	// ...
```

</details>

<details><summary>Logs</summary>

```
DEBUG: GitLab API 404 (repository=group1/group2/group3/my-app, branch=renovate/...)
       "url": "https://gitlab.example.com/api/v4/projects/somePath%2Fgroup1%2Fgroup2%2Fmy-dependency/repository/tree?per_page=100"
DEBUG: Error 404 getting changelog md (repository=somePath/group1/group2/my-dependency, branch=renovate/...)
       "type": "gitlab",
       "apiBaseUrl": "https://gitlab.example.com/api/v4/"
DEBUG: GET https://gitlab.example.com/api/v4/projects/somePath%2Fgroup1%2Fgroup2%2Fmy-dependency/releases?per_page=100 = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=13) (repository=group1/group2/group3/my-app, branch=renovate/...)
```

</details> 

<details><summary>- wrongUrl  + correctUrl</summary>

```diff
- https://gitlab.example.com/api/v4/projects/somePath%2Fgroup1%2Fgroup2%2Fmy-dependency/releases?per_page=100
+ https://gitlab.example.com/somePath/api/v4/projects/group1%2Fgroup2%2Fmy-dependency/releases?per_page=100
```

</details> 

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
What is the `packageFiles with updates` message for this dependency in debug logs?

**Reply by MalteHei** [reply]
<details><summary>packageFiles with updates</summary>

```
DEBUG: packageFiles with updates (repository=group1/group2/group3/my-app, baseBranch=main)
       "config": {
         "gitlabci-include": [
           {
             "deps": [
               {
                 "datasource": "gitlab-tags",
                 "depName": "group1/group2/my-dependency",
                 "depType": "repository",
                 "currentValue": "1.4.9",
                 "registryUrls": ["https://gitlab.example.com/somePath"],
                 "updates": [...],
                 "packageName": "group1/group2/my-dependency",
                 "versioning": "semver",
                 "warnings": [],
                 "sourceUrl": "https://gitlab.example.com/somePath/group1/group2/my-dependency",
                 "registryUrl": "https://gitlab.example.com/somePath",
                 "currentVersion": "1.4.9",
                 "isSingleVersion": true,
                 "fixedVersion": "1.4.9"
               }
             ],
             "packageFile": ".gitlab-ci.yml"
           }
         ],
         "maven": [...]
       }
```

</details>

**Reply by rarkins** [reply]
Do those URLs look ok? Is it the conversion from `https://gitlab.example.com/somePath/group1/group2/my-dependency` to `https://gitlab.example.com/api/v4/projects/somePath%2Fgroup1%2Fgroup2%2Fmy-dependency/releases?per_page=100` where the problem likely occurs?

**Reply by MalteHei** [reply]
That conversion is causing the problem, yes.

`somePath` is not a group in our GitLab instance.

As defined in our `config.js`, our API is reachable at `gitlab.example.com/somePath/api/v4/`, not `gitlab.example.com/api/v4/`.

**Reply by rarkins** [reply]
In that case we probably need to pass both the `sourceUrl` and `registryUrl` to the function which converts `sourceUrl` to the API endpoint

**Reply by rarkins** [reply, marked as answer]
I've created a bug issue here: #24512
